File: frontend/lib/schemas.ts
import { z } from "zod";

export const documentSchema = z.object({
  id: z.string(),
  name: z.string(),
  file_type: z.string(),
  status: z.enum(["queued", "processing", "ready", "failed"]),
  chunk_count: z.number(),
  page_count: z.number(),
  uploaded_at: z.string(),
  summary: z.string().nullable().optional()
});

export const citationSchema = z.object({
  document_id: z.string(),
  document_name: z.string(),
  page: z.number(),
  paragraph: z.number(),
  text: z.string(),
  score: z.number()
});

export const uploadResponseSchema = z.object({
  document: documentSchema,
  chunks_preview: z.array(citationSchema)
});

export const queryResponseSchema = z.object({
  answer: z.string(),
  citations: z.array(citationSchema),
  latency_ms: z.number(),
  provider: z.string()
});

export const questionSchema = z.object({
  question: z.string().trim().min(2, "Ask a fuller question.").max(2000)
});

export type DocumentMeta = z.infer<typeof documentSchema>;
export type Citation = z.infer<typeof citationSchema>;
export type UploadResponse = z.infer<typeof uploadResponseSchema>;
export type QueryResponse = z.infer<typeof queryResponseSchema>;

export type QueryMode = "single" | "compare";

export type ChatMessage = {
  id: string;
  role: "user" | "assistant";
  content: string;
  createdAt: string;
  citations?: Citation[];
  pending?: boolean;
};

File: backend/app/models/schemas.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class DocumentMeta(BaseModel):
    id: str
    name: str
    file_type: str
    status: Literal["queued", "processing", "ready", "failed"]
    chunk_count: int = 0
    page_count: int = 0
    uploaded_at: datetime
    summary: str | None = None


class Chunk(BaseModel):
    id: str
    document_id: str
    document_name: str
    text: str
    page: int
    paragraph: int
    embedding: list[float] = Field(default_factory=list)


class Citation(BaseModel):
    document_id: str
    document_name: str
    page: int
    paragraph: int
    text: str
    score: float


class QueryRequest(BaseModel):
    question: str = Field(min_length=2, max_length=2000)
    document_ids: list[str] = Field(default_factory=list)
    mode: Literal["single", "compare"] = "single"
    top_k: int = Field(default=5, ge=1, le=12)


class QueryResponse(BaseModel):
    answer: str
    citations: list[Citation]
    latency_ms: int
    provider: str


class StreamEvent(BaseModel):
    type: Literal["token", "citations", "done", "error"]
    value: str | list[Citation] | None = None


class UploadResponse(BaseModel):
    document: DocumentMeta
    chunks_preview: list[Citation]


class DeleteResponse(BaseModel):
    deleted: bool
    document_id: str
